components/layout_2020/display_list/background.rs

```diff
@@ -0,0 +1,250 @@
+                width = intrinsic.width;
+                height = intrinsic.height;
+            }
+
+            match (width, height) {
+                (Some(w), Some(h)) => units::LayoutSize::new(w.px(), h.px()),
+                (Some(w), None) => {
+                    let h = if let Some(intrinsic_ratio) = intrinsic.ratio {
+                        w / intrinsic_ratio
+                    } else if let Some(intrinsic_height) = intrinsic.height {
+                        intrinsic_height
+                    } else {
+                        // Treated as 100%
+                        Length::new(positioning_area.size.height)
+                    };
+                    units::LayoutSize::new(w.px(), h.px())
+                },
+                (None, Some(h)) => {
+                    let w = if let Some(intrinsic_ratio) = intrinsic.ratio {
+                        h * intrinsic_ratio
+                    } else if let Some(intrinsic_width) = intrinsic.width {
+                        intrinsic_width
+                    } else {
+                        // Treated as 100%
+                        Length::new(positioning_area.size.width)
+                    };
+                    units::LayoutSize::new(w.px(), h.px())
+                },
+                // Both comptued values were 'auto', and neither intrinsic size is present
+                (None, None) => size_contain_or_cover(ContainOrCover::Contain),
+            }
+        },
+    };
+
+    if tile_size.width == 0.0 || tile_size.height == 0.0 {
+        return None;
+    }
+
+    let mut tile_spacing = units::LayoutSize::zero();
+    let RepeatXY(repeat_x, repeat_y) = *get_cyclic(&b.background_repeat.0, layer_index);
+    let result_x = layout_1d(
+        &mut tile_size.width,
+        &mut tile_spacing.width,
+        repeat_x,
+        get_cyclic(&b.background_position_x.0, layer_index),
+        painting_area.origin.x - positioning_area.origin.x,
+        painting_area.size.width,
+        positioning_area.size.width,
+    );
+    let result_y = layout_1d(
+        &mut tile_size.height,
+        &mut tile_spacing.height,
+        repeat_y,
+        get_cyclic(&b.background_position_y.0, layer_index),
+        painting_area.origin.y - positioning_area.origin.y,
+        painting_area.size.height,
+        positioning_area.size.height,
+    );
+    let bounds = units::LayoutRect::new(
+        positioning_area.origin + Vector2D::new(result_x.bounds_origin, result_y.bounds_origin),
+        Size2D::new(result_x.bounds_size, result_y.bounds_size),
+    );
+
+    Some(Layer {
+        common,
+        bounds,
+        tile_size,
+        tile_spacing,
+        repeat: result_x.repeat || result_y.repeat,
+    })
+}
+
+/// Abstract over the horizontal or vertical dimension
+/// Coordinates (0, 0) for the purpose of this function are the positioning area’s origin.
+fn layout_1d(
+    tile_size: &mut f32,
+    tile_spacing: &mut f32,
+    mut repeat: Repeat,
```

Comment:
Any reason to have input/output parameters and a return struct here instead of simply returning these values in `Layout1DResult`?

Reply:
I’ve done so for `tile_spacing` but kept `tile_size` since it already has a non-trivial value before this call and is only sometimes modified.